portfolio/src/main/java/com/google/sps/PuzzleDao.java

```diff
@@ -0,0 +1,63 @@
+package com.google.sps;
+
+import com.google.cloud.datastore.Datastore;
+import com.google.cloud.datastore.DatastoreOptions;
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.FullEntity;
+import com.google.cloud.datastore.KeyFactory;
+
+/** Puzzle DAO implementation*/
+public class PuzzleDao implements Dao<Puzzle> {
+
+    private Datastore datastore;
+    private KeyFactory keyFactory;
+
+    public PuzzleDao() {
+      datastore = DatastoreOptions.getDefaultInstance().getService();
+      keyFactory = datastore.newKeyFactory().setKind("Puzzle");
+    }
+
+    @Override
+    public void create(Puzzle puzzle) {
+      //saves puzzle object on datastore
+      FullEntity taskEntity =
+        Entity.newBuilder(keyFactory.newKey())
+          .set("image", puzzle.getImage())
+          .build();
+      datastore.put(taskEntity);
+    }
+
+    @Override
+    public void update(Puzzle puzzle) {
+      //converts the id of the object into a Long.
+      Long puzzleId = Long.parseLong(puzzle.getPuzzleId());
+
+      //updates the entity with the data contained in "puzzle", 
+      // except the key which is used to search for the entity.
+      Entity task = 
+        Entity.newBuilder(datastore.get(keyFactory.newKey(puzzleId)))
+          .set("image", puzzle.getImage())
+          .build();
+      datastore.update(task);
+    }
+
+    @Override
+    public Puzzle read(String puzzleId) {
+      //reads the entity whose id match "puzzleID"
+      Entity entity = datastore.get(keyFactory.newKey(Long.parseLong(puzzleId)));
```

Comment:
```java
keyFactory.newKey(Long.parseLong(puzzleId))
```
Looks like this snippet of code is used often in this file.  I'd recommend refactoring it into a method so it can be reusable.

Reply:
I've also noticed that `entity.getKey().getId().toString()` is also used pretty often, should I put that into a function?